File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        unordered_map<string, string> map;
        for(auto& i : knowledge)
            map[i[0]] = i[1];
        
        string ans = "";
        
        string key = "";
        bool flag = false;
        
        for(int i=0; i<s.size(); i++) {
            if(!flag && s[i] != '(' && s[i] != ')')
                ans += s[i];
            
            else if(!flag && s[i] == '(')
                flag = true;
            
            else if(flag && s[i] != ')')
                key += s[i];
            
            else if(flag && s[i] == ')') {
                // evaluate key
                if(map.find(key) == map.end())
                    ans += "?"; // key not found
                else
                    ans += map[key];
                
                key = "";
                flag = false;
            }
        }
        
        return ans;
    }
};

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        HashMap<String, String> map = new HashMap<>();

        for (List<String> pair : knowledge) {
            map.put(pair.get(0), pair.get(1));
        }

        StringBuilder ans = new StringBuilder();
        int n = s.length();
        int i = 0;

        while (i < n) {
            char ch = s.charAt(i);
            StringBuilder temp = new StringBuilder();
            if (ch == '(') {
                i++;
                while (s.charAt(i) != ')') {
                    temp.append(s.charAt(i));
                    i++;
                }
                ans.append(map.getOrDefault(temp.toString(), "?"));
            } else {
                ans.append(ch);
            }
            i++;
        }

        return ans.toString();
    }
}